cdap-data-fabric/src/main/java/co/cask/cdap/spi/data/sql/jdbc/DataSourceInstantiator.java

```diff
@@ -64,8 +64,15 @@ public DataSource get() {
     if (dataSource != null) {
       return dataSource;
     }
+    return constructDataSource();
+  }
 
-    if (cConf.get(Constants.Dataset.DATA_STORAGE_IMPLEMENTATION).equals(Constants.Dataset.DATA_STORAGE_NOSQL)) {
+  private synchronized DataSource constructDataSource() {
+    // this is needed to prevent recreation of the DataSource if the get() method is called async
```

Comment:
`this is needed to prevent recreation of the DataSource if the get() method is called concurrently`